runtime/stackmap/fixreturns.c

```diff
@@ -134,25 +134,51 @@ getReturnBytecode(J9ROMClass * romClass, J9ROMMethod * romMethod, UDATA * return
 	*returnSlots = 0;
 	sigChar = sigData[sigLength - 1];
 
+	if ('[' == sigData[sigLength - 2]) {
+		sigChar = '[';
+	}
+
 	if (sigChar != 'V') {
 		*returnSlots = 1;
 		if (sigChar == 'J' || sigChar == 'D') {
-			if (sigData[sigLength - 2] != '[') {
-				*returnSlots = 2;
-			}
+			*returnSlots = 2;
 		}
 	}
 
 	/* Determine the correct return bytecode to insert */
 
-	returnBytecode = JBreturn0 + (U_8) *returnSlots;
 
 	if ((J9UTF8_DATA(name)[0] == '<') && (J9UTF8_DATA(name)[1] == 'i')) {
-		returnBytecode = JBreturnFromConstructor;
-	} else if (romMethod->modifiers & J9AccSynchronized) {
-		returnBytecode = JBsyncReturn0 + (U_8) *returnSlots;
+		return JBreturnFromConstructor;
 	}
 
+	/* bool, byte, char, and short need special treatment since they need to be truncated before return */
+	switch(sigChar){
+		case 'Z':
+			returnBytecode = JBreturnZ;
+			break;
+		case 'B':
+			returnBytecode = JBreturnB;
+			break;
+		case 'C':
+			returnBytecode = JBreturnC;
+			break;
+		case 'S':
+			returnBytecode = JBreturnS;
+			break;
+		default:
+			if (romMethod->modifiers & J9AccSynchronized) {
+				return JBsyncReturn0 + (U_8) *returnSlots;
```

Comment:
This change introduces multiple return points - can we keep it to a single return?